### DISCUSSION TITLE
Virtiofs support

### DISCUSSION THREAD

**Discussion by xanderio** [seaweedfs/seaweedfs#3434] [Ideas]
## Problem

One possible workload of seaweedfs would be to provide persistent storage to a vm. There are currently to options to do this: 
1. Mounting inside the VM via FUSE, this has the problem that you need credentials inside the VM.
2. Mounting via FUSE on the host and then share this mount to the VM, with e.g. virtiofsd. This has the drawback that any I/O needs to go throw two separate programs.

## Proposal

seaweedfs gains a new capability to directly provide a virtiofs socket. 

## Futher reading
https://virtio-fs.gitlab.io/

**Comment by chrislusf** [comment]
That is a good possibility.

virtiofsd seems ok for now.

I wonder what is the way to provide storage to a VM with the best performance?

**Reply by xanderio** [reply]
As far as I can tell virtiofs seams to be the fastest.

**Reply by chrislusf** [reply]
I read through the docs and think virtiofs is nice. I have basic Rust knowledge, and would need some help to have a basic prototype.

**Reply by xanderio** [reply]
I think I should be able to help you with this, but I'm not sure how Rust comes into the equation. 
Are you planing to create a separate Rust based solution for this, or do you want this functionality to build in included in the main go project?

**Reply by chrislusf** [reply]
I only saw a rust implementation for virtiofsd. Not sure what's the best way to do this.

**Reply by xanderio** [reply]
I try to build a prototype using [fuse-backend-rs](https://lib.rs/crates/fuse-backend-rs) and see how far I get :)

**Reply by chrislusf** [reply]
I was looking at https://gitlab.com/virtio-fs/virtiofsd Seems it is more "official".

**Reply by chrislusf** [reply]
I have compiled https://gitlab.com/virtio-fs/virtiofsd code locally. Not exactly sure what is the most up-to-date way to use it. Do I need to use Kata? Seems docker has support for virtiofs too.
